src/cgutils.cpp

```diff
@@ -1148,6 +1198,8 @@ static jl_cgval_t typed_load(Value *ptr, Value *idx_0based, jl_value_t *jltype,
     //else {
         Instruction *load = builder.CreateAlignedLoad(data, isboxed ?
             alignment : julia_alignment(data, jltype, alignment), false);
+        if (isboxed)
+            load = maybe_mark_load_dereferenceable(load, true, jltype);
```

Comment:
I don't think we can guarantee that we will only ever form valid pointers. For example, someone's trying to add a PR right now (#21263) that appears to depend on this not being here.